src/corefx/System.Globalization.Native/casing.cpp

```diff
@@ -13,87 +13,140 @@
-    // (The loop here has been manually cloned for each of the four cases, rather
-    // than having a
-    // single loop that internally branched based on bToUpper and bTurkishCasing,
-    // as the compiler
-    // wasn't doing that optimization, and it results in an ~15-20% perf
-    // improvement on longer strings.)
+	UBool isError = FALSE;
+	int32_t srcIdx = 0, dstIdx = 0;
+	UChar32 srcCodepoint, dstCodepoint;
 
-    if (bToUpper)
-    {
-        if (bTurkishCasing)
-        {
-            // bToUpper && bTurkishCasing
-            while (srcIdx < cwSrcLength)
-            {
-                // In turkish casing, LATIN SMALL LETTER I (U+0069) upper cases to LATIN
-                // CAPITAL LETTER I WITH DOT ABOVE (U+0130).
-                U16_NEXT(lpSrc, srcIdx, cwSrcLength, srcCodepoint);
-                dstCodepoint = ((srcCodepoint == (UChar32)0x0069) ? (UChar32)0x0130 : u_toupper(srcCodepoint));
-                U16_APPEND(lpDst, dstIdx, cwDstLength, dstCodepoint, isError);
-                assert(isError == FALSE && srcIdx == dstIdx);
-            }
-        }
-        else
-        {
-            // bToUpper && !bTurkishCasing
-            while (srcIdx < cwSrcLength)
-            {
-                U16_NEXT(lpSrc, srcIdx, cwSrcLength, srcCodepoint);
-                dstCodepoint = u_toupper(srcCodepoint);
-                U16_APPEND(lpDst, dstIdx, cwDstLength, dstCodepoint, isError);
-                assert(isError == FALSE && srcIdx == dstIdx);
-            }
-        }
-    }
-    else
-    {
-        if (bTurkishCasing)
-        {
-            // !bToUpper && bTurkishCasing
-            while (srcIdx < cwSrcLength)
-            {
-                // In turkish casing, LATIN CAPITAL LETTER I (U+0049) lower cases to
-                // LATIN SMALL LETTER DOTLESS I (U+0131).
-                U16_NEXT(lpSrc, srcIdx, cwSrcLength, srcCodepoint);
-                dstCodepoint = ((srcCodepoint == (UChar32)0x0049) ? (UChar32)0x0131 : u_tolower(srcCodepoint));
-                U16_APPEND(lpDst, dstIdx, cwDstLength, dstCodepoint, isError);
-                assert(isError == FALSE && srcIdx == dstIdx);
-            }
-        }
-        else
-        {
-            // !bToUpper && !bTurkishCasing
-            while (srcIdx < cwSrcLength)
-            {
-                U16_NEXT(lpSrc, srcIdx, cwSrcLength, srcCodepoint);
-                dstCodepoint = u_tolower(srcCodepoint);
-                U16_APPEND(lpDst, dstIdx, cwDstLength, dstCodepoint, isError);
-                assert(isError == FALSE && srcIdx == dstIdx);
-            }
-        }
-    }
+	if (bToUpper)
+	{
+		while (srcIdx < cwSrcLength)
+		{
+			// In turkish casing, LATIN SMALL LETTER I (U+0069) upper cases to LATIN
+			// CAPITAL LETTER I WITH DOT ABOVE (U+0130).
+			U16_NEXT(lpSrc, srcIdx, cwSrcLength, srcCodepoint);
+			dstCodepoint = ((srcCodepoint == (UChar32)0x0069) ? (UChar32)0x0130 : u_toupper(srcCodepoint));
```

Comment:
With this change, the TestToLowerToUpper_TurkishI test at https://github.com/dotnet/corefx/pull/3976/files#diff-a56516550b87790fea0341b950837fbcR1462 passes.  Does that answer your question?  If not, can you clarify?